### PR TITLE
Support `cl make --store` to realize the functionality of `cl replicate`

### LINKED ISSUE #3922: Implement "cl replicate" CLI command
Design doc: https://docs.google.com/document/d/1TOsO4gnxUm1XMmlGUb7BJ4Z6y_olFbv9APbx0-IEy6M/edit#

cl replicate [bundle uuid] --store [store name]
Replicates the contents of the specified bundle to the specified bundle store.

### PR DESCRIPTION
**Pull request by wwwjn**
### Reasons for making this change

Extend `cl make` to have `cl  replicate` function, which let user move a bundle from one bundle storage to another bundle storage.

### Related issues

Fix issue #3922 

### Checklist

* [ ] I've added a screenshot of the changes, if this is a frontend change
* [x] I've added and/or updated tests, if this is a backend change
* [x] I've run the [pre-commit.sh](https://github.com/codalab/codalab-worksheets/blob/master/pre-commit.sh) script
* [x] I've updated [docs](https://github.com/codalab/codalab-worksheets/tree/master/docs), if needed

### CODE REVIEW

**Comment by wwwjn** [comment]
TODO: add more tests for "cl make"

File: codalab/bundles/make_bundle.py [hunk 1]
```diff
@@ -20,6 +20,16 @@ class MakeBundle(DerivedBundle):
             generated=True,
         )
     )
+    METADATA_SPECS.append(  # support `--store` command, set the
```
Comment (epicfaace) [inline codalab/bundles/make_bundle.py]:
```suggestion
    METADATA_SPECS.append(
```

Remove this comment?

File: codalab/lib/bundle_cli.py [hunk 2]
```diff
@@ -1718,6 +1719,30 @@ def do_make_command(self, args):
             params={'worksheet': worksheet_uuid},
         )
 
+        # Add new bundle's location. If user specify the storage using `--store`, the bundle will be added to that storage.
```
Comment (epicfaace) [inline codalab/lib/bundle_cli.py:1709]:
```suggestion
        # Add new bundle's location. If the user specifies the bundle store using `--store`, the bundle will be added to that store.
```

File: codalab/lib/bundle_cli.py [hunk 3]
```diff
@@ -1718,6 +1719,30 @@ def do_make_command(self, args):
...
+        # Otherwise, the new MakeBundle will be added to default storage, which is set by the rest server.
```
Comment (epicfaace) [inline codalab/lib/bundle_cli.py:1710]:
```suggestion
        # Otherwise, the new MakeBundle will be added to a default store set by the rest server.
```

File: codalab/lib/bundle_cli.py [hunk 4]
```diff
@@ -1718,6 +1719,30 @@ def do_make_command(self, args):
...
+                        'name': destination_bundle_store,
+                        'include': ['uuid', 'storage_type', 'url'],
+                    },
+                )
+                bundle_store_uuid = storage_info['uuid']
+                if storage_info['storage_type'] in (StorageType.DISK_STORAGE.value,):
+                    need_bypass = False  # The user specify --store to upload to disk storage
+
+            # 2) Add bundle to destination bundle storage
+            params = {'need_bypass': need_bypass}  # is_dir: ???
```
Comment (epicfaace) [inline codalab/lib/bundle_cli.py]: I think is_dir will be True if cl make was input with multiple arguments. If cl make was input with a single argument that points to a file (not a directory), is_dir will be False. See https://codalab-worksheets.readthedocs.io/en/latest/CLI-Reference/#make for the two scenarios. adding @percyliang to verify though

File: codalab/lib/bundle_cli.py [hunk 5]
```diff
@@ -1718,6 +1719,30 @@ def do_make_command(self, args):
...
+                storage_info = client.fetch_one(
+                    'bundle_stores',
+                    params={
+                        'name': destination_bundle_store,
+                        'include': ['uuid', 'storage_type', 'url'],
+                    },
+                )
+                bundle_store_uuid = storage_info['uuid']
+                if storage_info['storage_type'] in (StorageType.DISK_STORAGE.value,):
+                    need_bypass = False  # The user specify --store to upload to disk storage
```
Comment (epicfaace) [inline codalab/lib/bundle_cli.py]: What does this do? `need_bypass` was already set to False.

File: codalab/lib/bundle_cli.py [hunk 6]
```diff
@@ -1718,6 +1719,30 @@ def do_make_command(self, args):
...
+
+            # 2) Add bundle to destination bundle storage
+            params = {'need_bypass': need_bypass}  # is_dir: ???
+            client.add_bundle_location(new_bundle['uuid'], bundle_store_uuid, params)
+
```
Comment (epicfaace) [inline codalab/lib/bundle_cli.py:1729]: Should we add code that handles the case when `destination_bundle_store` is not specified (and needs to be set to a default value)?

File: codalab/rest/bundles.py [hunk 7]
```diff
@@ -259,6 +259,7 @@ def _create_bundles():
         .load(request.json)
         .data
     )
+    logging.info("POST /Bundle is called....")
```
Comment (epicfaace) [inline codalab/rest/bundles.py]: remove

File: codalab/rest/bundles.py [hunk 8]
```diff
@@ -467,12 +468,12 @@ def _add_bundle_location(bundle_uuid: str):
     """
     check_bundles_have_all_permission(local.model, request.user, [bundle_uuid])
     need_bypass = query_get_bool('need_bypass', default=False)
-    is_dir = query_get_bool('is_dir', default=False)
+    is_dir = query_get_bool('is_dir', default=None)
```
Comment (epicfaace) [inline codalab/rest/bundles.py:470]: Why this change?

File: codalab/server/bundle_manager.py [hunk 9]
```diff
@@ -226,9 +226,11 @@ def _is_making_bundle(self, uuid):
 
     def _make_bundle(self, bundle):
         try:
+            # If it's link bundle, the actual storage place will be in the link_url
```
Comment (epicfaace) [inline codalab/server/bundle_manager.py]:
```suggestion
            # For linked bundles, the actual storage place will be in the link_url
```

File: codalab/server/bundle_manager.py [hunk 10]
```diff
@@ -226,9 +226,11 @@ def _is_making_bundle(self, uuid):
...
             bundle_link_url = getattr(bundle.metadata, "link_url", None)
             bundle_location = bundle_link_url or self._bundle_store.get_bundle_location(bundle.uuid)
 
+            # Here the path might be a blob storage or
```
Comment (epicfaace) [inline codalab/server/bundle_manager.py]: finish comment?

File: codalab/server/bundle_manager.py [hunk 11]
```diff
@@ -263,7 +265,10 @@ def _make_bundle(self, bundle):
 
                     # If source path is on Azure Blob Storage, we should download it to a temporary local directory first.
                     if parse_linked_bundle_url(dependency_path).uses_beam:
-                        dependency_path = os.path.join(tempdir, dep.parent_uuid)
+                        if dep.child_path != "":
```
Comment (epicfaace) [inline codalab/server/bundle_manager.py:267]: What does this code do?

File: tests/cli/test_cli.py [hunk 12]
```diff
@@ -1372,6 +1372,46 @@ def test_make(ctx):
     _run_command([cl, 'rm', '--force', uuid2])  # force the deletion
     _run_command([cl, 'rm', '-r', uuid1])  # delete things downstream
 
+    # test using make to replicate bundles between storages
```
Comment (epicfaace) [inline tests/cli/test_cli.py]:
```suggestion
    # test using make to replicate bundles between bundle stores
```

File: tests/cli/test_cli.py [hunk 13]
```diff
@@ -1372,6 +1372,46 @@ def test_make(ctx):
...
+                cl,
+                "store",
+                "add",
+                "--name",
+                bundle_store_name,
+                '--url',
+                'azfs://devstoreaccount1/bundles',
+            ]
+        )
+        # 1) blob storage -> local filesystem
```
Comment (epicfaace) [inline tests/cli/test_cli.py]: what do the 1, 2, 3 refer to? the items in parent_child_store? if so, we might want to move those comments to be inline with the array item declarations

**Review by epicfaace — CHANGES_REQUESTED** [review]
Looks great so far!

File: codalab/lib/bundle_cli.py (hunk 5 above)
Comment (wwwjn) [reply codalab/lib/bundle_cli.py]: Thanks for reminding me this! Already remove this part.

File: codalab/lib/bundle_cli.py (hunk 6 above)
Comment (wwwjn) [reply codalab/lib/bundle_cli.py:1729]: Great question! The client.add_bundle_locations function will trigger the client side to handle this. This means if the user does not specify the `destination_bundle_store`, rest server will add the new bundle to default storage

File: codalab/rest/bundles.py (hunk 8 above)
Comment (wwwjn) [reply codalab/rest/bundles.py:470]: Originally we only use this API for `cl upload`, where by default the `is_dir` is set according to the actual file/ folder the user uploaded.
Now we reuse the API for `cl make [bundle_id]  --store`, the user might make multiple bundles into 1 bundle, or make a bundle contains a folder. We can not know `is_dir` in advance, so by default we set it to `None`.

File: codalab/server/bundle_manager.py (hunk 11 above)
Comment (wwwjn) [reply codalab/server/bundle_manager.py:267]: The  `dep.child_path` is the key specified by user, eg `cl make key1:[bundle_id1] --store azure`.  I want to use the key as the new sub-folder name if the user are making multiple bundle into one folder.

File: codalab/lib/bundle_cli.py (hunk 4 above)
Comment (wwwjn) [reply codalab/lib/bundle_cli.py]: This is an old commit and I already remove this. In `cl make`, we first set `is_dir = None` here. And then in bundle manager, the function `_make_bundle` will check whether there are multiple input bundle, or the input bundle is file or directory, and update `is_dir` field.

File: codalab/lib/upload_manager.py [hunk 14]
```diff
@@ -133,8 +133,6 @@ def _update_and_get_bundle_location(self, bundle: Bundle, is_directory: bool) ->
             str: Bundle location.
         """
         if self.destination_bundle_store is not None:
-            # In this case, we are using the new BundleStore / BundleLocation model to track the bundle location.
```
Comment (epicfaace) [inline codalab/lib/upload_manager.py:142]: keep these comments?

File: codalab/rest/bundles.py (hunk 8 above)
Comment (epicfaace) [reply codalab/rest/bundles.py:470]: But how do we validate `is_dir` then, particularly if the user is using `cl upload`? Shouldn't it be ultimately set to True or False once we create the new bundle?

File: codalab/server/bundle_manager.py [hunk 15]
```diff
@@ -226,9 +226,11 @@ def _is_making_bundle(self, uuid):
 
     def _make_bundle(self, bundle):
         try:
+            # For linked bundles, the actual storage place will be in the link_url
```
Comment (epicfaace) [inline codalab/server/bundle_manager.py]: do we need these two comments?

File: codalab/server/bundle_manager.py (hunk 11 above)
Comment (epicfaace) [reply codalab/server/bundle_manager.py:267]: Let's replace this whole if statement with

```
dependency_path = os.path.join(tempdir, dep.child_path or dep.parent_uuid)
```

And add a comment explaining when we would use dep.child_path

File: codalab/server/bundle_manager.py [hunk 16]
```diff
@@ -281,17 +286,42 @@ def _make_bundle(self, bundle):
                             with open(dependency_path, 'wb') as f:
                                 shutil.copyfileobj(fileobj, f)
 
+                    # If source is local file system and destination is blob storage:
```
Comment (epicfaace) [inline codalab/server/bundle_manager.py]:
```suggestion
                    # If source is local file system and destination is blob storage,
```

File: codalab/server/bundle_manager.py [hunk 17]
```diff
@@ -281,17 +286,42 @@ def _make_bundle(self, bundle):
...
+                    # need to copy everything into a temp folder and upload together
+                    elif parse_linked_bundle_url(path).uses_beam:
+                        if dep.child_path != "":
+                            tempdir_dependency_path = os.path.join(tempdir, dep.child_path)
```
Comment (epicfaace) [inline codalab/server/bundle_manager.py]: Make this into an one-liner as before (`dep.child_path or dep.parent_uuid`)

File: codalab/server/bundle_manager.py [hunk 18]
```diff
@@ -281,17 +286,42 @@ def _make_bundle(self, bundle):
...
+                        dependency_path = tempdir_dependency_path
                     deps.append((dependency_path, child_path))
+                remove_path(path)  # delete the original bundle path
+
+                # Upload to destination bundle storage
+                if parse_linked_bundle_url(path).uses_beam:  # The destination is using blob storage
+                    source_fileobj = zip_util.tar_gzip_directory(
+                        tempdir
+                    )  # pack all the files in temp folder to a fileobj
+                    source_filename = "MakeBundle.tar.gz"  # TODO(Jiani): Use the original filename
```
Comment (epicfaace) [inline codalab/server/bundle_manager.py:308]: should we fix this for this PR? and does this file name even matter / does it get saved in the output bundle?

File: codalab/server/bundle_manager.py [hunk 19]
```diff
@@ -281,17 +286,42 @@ def _make_bundle(self, bundle):
...
-                remove_path(path)
-
-                if len(deps) == 1 and deps[0][1] == path:
-                    path_util.copy(deps[0][0], path, follow_symlinks=False)
-                else:
-                    os.mkdir(path)
-                    for dependency_path, child_path in deps:
-                        path_util.copy(dependency_path, child_path, follow_symlinks=False)
-
+            bundle_location = bundle_link_url or self._bundle_store.get_bundle_location(bundle.uuid)
```
Comment (epicfaace) [inline codalab/server/bundle_manager.py:325]: why do we need this? It wasn't there before I think

File: codalab/server/bundle_manager.py (hunk 18 above)
Comment (wwwjn) [reply codalab/server/bundle_manager.py:308]: This filename ("MakeBundle.tar.gz") will not be save in the output bundle.  However, the output bundle will automatically be named as `make-[source bundle id]`. I will remove this comment. Thanks Ashwin!

File: codalab/lib/bundle_cli.py [hunk 20]
```diff
@@ -1706,6 +1706,27 @@ def do_make_command(self, args):
...
 
+        # Add new bundle's location. If user specify the storage using `--store`, the bundle will be added to that storage.
+        # Otherwise, the new MakeBundle will be added to default storage, which is set by the rest server.
+        destination_bundle_store = metadata.get('store')
+        if destination_bundle_store is not None:
+            need_bypass = False
+            bundle_store_uuid = None
+
+            # 1) Read destination store from --store if user has specified it
+            if destination_bundle_store is not None and destination_bundle_store != '':
```
Comment (epicfaace) [inline codalab/lib/bundle_cli.py]:
```suggestion
            if destination_bundle_store:
```

File: codalab/lib/bundle_cli.py [hunk 21]
```diff
@@ -1706,6 +1706,27 @@ def do_make_command(self, args):
             params={'worksheet': worksheet_uuid},
         )
 
+        # Add new bundle's location. If user specify the storage using `--store`, the bundle will be added to that storage.
+        # Otherwise, the new MakeBundle will be added to default storage, which is set by the rest server.
+        destination_bundle_store = metadata.get('store')
+        if destination_bundle_store is not None:
+            need_bypass = False
```
Comment (epicfaace) [inline codalab/lib/bundle_cli.py]: Remove this and just set need_bypass later
```suggestion
```

File: codalab/lib/bundle_cli.py [hunk 22]
```diff
@@ -1706,6 +1706,27 @@ def do_make_command(self, args):
...
+                storage_info = client.fetch_one(
+                    'bundle_stores',
+                    params={
+                        'name': destination_bundle_store,
+                        'include': ['uuid', 'storage_type', 'url'],
+                    },
+                )
+                bundle_store_uuid = storage_info['uuid']
+
+            params = {'need_bypass': need_bypass}
```
Comment (epicfaace) [inline codalab/lib/bundle_cli.py]: Can you also add a comment for why we should not bypass server for make bundles?

```suggestion
            params = {'need_bypass': False}
```

File: codalab/rest/bundles.py (hunk 8 above)
Comment (epicfaace) [reply codalab/rest/bundles.py:470]: I see!

File: codalab/rest/bundles.py [hunk 23]
```diff
@@ -467,12 +467,12 @@ def _add_bundle_location(bundle_uuid: str):
...
```
Comment (epicfaace) [inline codalab/rest/bundles.py:470]: @wwwjn Can you add an assertion, before we save the bundle to the database, like `assert is_dir in (True, False)`? This way we can make sure it's never set to None.

File: codalab/server/bundle_manager.py (hunk 11 above)
Comment (epicfaace) [reply codalab/server/bundle_manager.py:267]: @wwwjn can you update this?

File: codalab/server/bundle_manager.py (hunk 18 above)
Comment (epicfaace) [reply codalab/server/bundle_manager.py:308]: @wwwjn can you remove the comment, and also add another comment noting the filename doesn't make a difference / isn't saved in the output?

File: codalab/server/bundle_manager.py (hunk 19 above)
Comment (epicfaace) [reply codalab/server/bundle_manager.py:325]: @wwwjn

**Review by epicfaace — APPROVED** [review]
